{"cells":[{"metadata":{},"cell_type":"markdown","source":"# Step 1: Upload data files: Click File:Open\n\nPress the Double Arrow above to Begin, then click red pop-up to confirm"},{"metadata":{"hide_input":false,"scrolled":true,"trusted":true},"cell_type":"code","source":"display(\"Loading Allen Transcriptomics Single-Cell RNAseq data, retreived May 2020. Be patient, it's loading!\")\n","execution_count":null,"outputs":[]},{"metadata":{"hide_input":false,"trusted":true},"cell_type":"code","source":"%%html\n<style>\n.output_wrapper, .output {\n    height:auto !important;\n    max-height:1000px;  /* your desired max-height here */\n}\n.output_scroll {\n    box-shadow:none !important;\n    webkit-box-shadow:none !important;\n}\n</style>","execution_count":1,"outputs":[{"output_type":"display_data","data":{"text/plain":"<IPython.core.display.HTML object>","text/html":"<style>\n.output_wrapper, .output {\n    height:auto !important;\n    max-height:1000px;  /* your desired max-height here */\n}\n.output_scroll {\n    box-shadow:none !important;\n    webkit-box-shadow:none !important;\n}\n</style>\n"},"metadata":{}}]},{"metadata":{"hide_input":false,"trusted":true},"cell_type":"code","source":"import matplotlib \nimport ipywidgets as widgets\nimport pandas as pd\nimport io\nimport seaborn as sns\nfrom IPython.display import display\nfrom ipywidgets import interact_manual, interactive,GridspecLayout, AppLayout, Button, Layout","execution_count":2,"outputs":[]},{"metadata":{"hide_input":false,"trusted":true},"cell_type":"code","source":"human_means = pd.read_csv(\"human_means.csv\", index_col=0)\nmeds = pd.read_csv(\"meds.csv\", index_col=0)   \n    \n\n# Get all gene and cell type names: HUMAN\nhuman_listOfColumnNames = list(human_means.columns.values)\nhuman_idx = list(human_means.index)\n# Get all gene and cell type: MOUSE\nmouse_listOfColumnNames = list(meds.columns.values)\nidx = list(meds.index)\n","execution_count":11,"outputs":[]},{"metadata":{"hide_input":false,"trusted":true},"cell_type":"code","source":"##define GUI\ntable_style = {'description_width': 'initial'}\nuseidx = ('choose species','then genes will appear')\nspacer = widgets.Label('')\n\nfrom threading import Timer\n\nfrom traitlets import traitlets\n\nclass LoadedButton(widgets.Button):\n    \"\"\"A button that can hold a value as a attribute.\"\"\"\n\n    def __init__(self, value=None, *args, **kwargs):\n        super(LoadedButton, self).__init__(*args, **kwargs)\n        # Create the value attribute.\n        self.add_traits(value=traitlets.Any(value))\n#####################################################        \ngenes = []\ngenetitle = widgets.Label('Select Genes')\ndropdown =  widgets.Combobox(\n    options= idx,\n    description='Gene',\n    placeholder='Type here',\n    layout={'width': '80%'})\n\ndef delay():\n    gene_button.button_style = ''\n    gene_button.disabled = False\n    \ndef choose_gene(change):\n    if len(genes)<=7:\n        gene_button.button_style = 'success'\n        gene_button.disabled = True\n        genes.append(dropdown.value)     \n        d = Timer(0.5,delay)\n        d.start()\n    if len(genes)>7:       \n        gene_button.button_style = 'warning'\n        gene_button.description = \"Max Genes Selected\"\n        gene_button.disabled = True\n        \ngene_button = widgets.Button(description = 'Enter Gene', button_style = '')\ngene_button.on_click(choose_gene)\n\nanalysislabel = widgets.Label('Graph Type')        \ndef on_exp_button_clicked(exp):\n    exp.value = \"exp\"\n    venn.disabled = True\n    exp.button_style = 'success'\n    graphtype = \"Exp\"\n\n\ndef on_venn_button_clicked(venn):\n    venn.value = \"venn\"\n    venn.button_style = 'success'\n    exp.disabled = True\n    graphtype = \"Venn\"\n    \nexp = LoadedButton(description=\"Expression Level\", value=\"not exp\")\nexp.on_click(on_exp_button_clicked)         \n\nvenn = LoadedButton(description=\"Venn Diagram\", value=\"not venn\")\nvenn.on_click(on_venn_button_clicked)\n\n\nchoicelabel = widgets.Label('Analyze species:')\nmsCells = widgets.Button(\n    description='Mouse',\n    disabled=False,\n    button_style='', # 'success', 'info', 'warning', 'danger' or ''\n    tooltip='Description',\n    width = '110%')\n\nhuCells = widgets.Button(\n    description='Human',\n    disabled=False,\n    button_style='', # 'success', 'info', 'warning', 'danger' or ''\n    tooltip='Description',\n    width = '110%')\n\n#####################################################\nmscelllabel = widgets.Label('Mouse Cell Clusters:')\nmsCells_drop =widgets.SelectMultiple(\n    options=mouse_listOfColumnNames,\n    disabled=True,\n    rows=17,\n    layout={'width': 'max-content'},\n    #style=table_style\n) \n\nhucelllabel = widgets.Label('Human Cell Clusters:')\nhuCells_drop =widgets.SelectMultiple(\n    options=human_listOfColumnNames,\n    disabled=True,\n    rows=17,\n    layout={'width': 'max-content'},\n    #style=table_style\n) \n\ndef choose_mouse(msCells):\n    huCells.disabled = True\n    msCells.button_style = 'success'\n    msCells_drop.disabled = False\n    huCells_drop.disabled = True \n    msCells.value = \"mouse\"\n\nmsCells.on_click(choose_mouse)\n\ndef choose_human(huCells):\n    huCells.button_style = 'success'\n    msCells.disabled = True\n    msCells_drop.disabled = True\n    huCells_drop.disabled = False\n    huCells.value = \"human\"\n    dropdown.options = human_idx\n\nhuCells.on_click(choose_human)\n\nmsCells = LoadedButton(description=\"Mouse\", value=\"not mouse\")\nmsCells.on_click(choose_mouse)   \n\nhuCells = LoadedButton(description=\"Human\", value=\"not human\")\nhuCells.on_click(choose_human) \ndropdown.observe(choose_human, 'change')\n########################################\n\nexcel_label = widgets.Text(\n    value='Name Excel:',\n    placeholder='Type something',\n    description='String:',\n    disabled=False\n)\n\ngo_button = widgets.Button(description = 'Produce Output', \n                           button_style = 'success', \n                            display='flex',\n                            flex_flow='column',\n                            align_items='stretch')\nreset_button = widgets.Button(description = 'Reset Options',\n                              button_style = 'warning',\n                            display='flex',\n                            flex_flow='column',\n                            align_items='stretch',\n                             width = 'auto')\n\nfrom ipywidgets import Widget\ncells =[]\ndef on_go_button_clicked(go_button):\n    excelname = excel_label.value\n    if huCells.value == 'human':\n        cells.append(huCells_drop.value)\n    else:\n        cells.append(msCells_drop.value)\n    if venn.value == \"venn\":\n        venngraph(genes, excelname)\n        \n    if exp.value == \"exp\":\n        expgraph(genes,cells,excelname)\n\ndef on_reset_button_clicked(reset_button):  \n    Widget.close_all()\n    \ngo_button.observe(huCells, 'value')\ngo_button.observe(venn, 'value')    \ngo_button.observe(exp, 'value')    \ngo_button.on_click(on_go_button_clicked)\nreset_button.on_click(on_reset_button_clicked)\n\n\n\nspacervenn = widgets.VBox([spacer,venn])\nspecieschoices = widgets.VBox([choicelabel,msCells, mscelllabel, msCells_drop])\nspacerhuman = widgets.VBox([spacer, huCells,hucelllabel, huCells_drop])\ngenesbox = widgets.VBox([genetitle,dropdown, gene_button])\nbuttons = widgets.HBox([go_button, spacer, reset_button])\ngraphchoices = widgets.VBox([analysislabel, exp, venn, spacer, spacer, genesbox,spacer, spacer, excel_label, buttons])\ngraphs2 = widgets.HBox([graphchoices])\n####################################################\ndisplay(\"..............\")\ndisplay(\"Choose your Species and Graph Type. For Expression level graphs, choose cell types to include. For Venn diagrams, all cell types will be included. Choose genes, pressing Enter after each choice. To produce graphs, press the green Produce Output button.\")","execution_count":null,"outputs":[]},{"metadata":{"hide_input":true,"trusted":true},"cell_type":"code","source":"%matplotlib notebook\n\ndef expgraph(genes, cells, excelname): \n    \n    if msCells.value == 'mouse':\n        bigframe = meds\n        cells = msCells_drop.value\n    if huCells.value == 'human':\n        bigframe = human_means\n        cells = huCells_drop.value\n\n    newdf = pd.DataFrame(columns = ['Cells','Genes','Expression'])\n    newdf['Cells'] = cells * len(genes)\n\n\n    genelist=[]\n    for i in range(len(genes)):\n        genea = [genes[i]]*len(cells)\n        genelist.append(genea)\n    flat_geneslist = [item for sublist in genelist for item in sublist]\n    newdf['Genes'] = flat_geneslist\n\n    expressions =[]\n    for i in range(len(genes)*len(cells)):\n        value1 = bigframe.loc[newdf['Genes'][i], newdf['Cells'][i]]\n        expressions.append(value1)\n    newdf['Expression'] = expressions\n\n    if len(newdf['Expression'])>0:\n        maxm = max(newdf[\"Expression\"])\n    \n\n    newdf.to_excel(excelname.value) \n    \n    sns.set_palette(\"colorblind\", 10)\n    g = sns.catplot(x=\"Genes\", y=\"Expression\", col=\"Cells\",\n                    data=newdf, \n                    saturation=1,\n                    kind=\"bar\", \n                    ci=None, \n                    aspect=.6, \n                    edgecolor=\"k\", \n                    linewidth=1\n                    )\n\n\n    (g.set_axis_labels(\"\", \"Expression\")\n          .set_xticklabels(genes)\n          .set_titles(\"{col_name}\")\n          .set(ylim=(0, (maxm + 2)))\n          .despine(left=True))  \n\n    for ax in g.axes.ravel():\n            ax.set_xticklabels(ax.get_xticklabels(), rotation=45, horizontalalignment='right')\n            for p in ax.patches:\n                ax.annotate('%.2f' % p.get_height(), (p.get_x() + p.get_width() / 2., p.get_height()), \n                ha='center', va='center', xytext=(0, 8), \n                textcoords='offset points')\n","execution_count":null,"outputs":[]},{"metadata":{"hide_input":false,"trusted":true},"cell_type":"code","source":"def venngraph(genes, excelname):\n        labels_yn = \"True\"\n        from matplotlib_venn import venn2\n        from matplotlib_venn import venn3\n        from matplotlib import pyplot as plt\n     \n        if msCells.value == 'mouse':\n            venny = meds.loc[genes]\n        if huCells.value == 'human':\n            venny = human_means.loc[genes]\n\n        venny1 = venny.loc[genes[0]]\n        venny1 = venny1[venny1 >= 0.5].index.to_list()\n\n        venny2 = venny.loc[genes[1]]\n        venny2 = venny2[venny2 >= 0.5].index.to_list()\n\n        v1only = list(set(venny1) - set(venny2))\n        v2only = list(set(venny2) - set(venny1))\n        overlap_ab = [value for value in venny1 if value in venny2] \n        \n        name1 = str(genes[0]) + \" only\"\n        name2 = str(genes[1]) + \" only\"\n        nameab = str(genes[0]) + \" and \" + str(genes[1])\n\n        \n        if len(genes) == 2:\n            if (len(venny1) >= 1) & (len(venny2) >= 1):\n                la,lb,lc = len(v1only),len(v2only),len(overlap_ab)\n                # now find the max\n                max_len = max(la,lb,lc)\n                if not max_len == la:\n                    v1only.extend(['']*(max_len-la))\n                if not max_len == lb:\n                    v2only.extend(['']*(max_len-lb))\n                if not max_len == lc:\n                    overlap_ab.extend(['']*(max_len-lc))\n                venn_df = pd.DataFrame({name1:v1only,name2:v2only,nameab:overlap_ab}) \n\n                labels = [\"Cells containing \" + genes[0],\"Cells containing \" + genes[1] ]\n                    \n                v = venn2([set(venny1), set(venny2)], set_labels = (genes[0], genes[1]))\n                 \n                venn_df.to_excel(excelname.value)\n                    #if labels_yn == \"True\":\n                        #v.get_label_by_id('10').set_text('\\n'.join(v1only))\n                        #v.get_label_by_id('01').set_text('\\n'.join(v2only))\n                        #v.get_label_by_id('11').set_text('\\n'.join(overlap_ab))\n                                \n            else:\n                print(\"No Overlap: At least one gene is not expressed over a value of 0.5 in any cell types:\", \n                      genes[0], \": expressed in \", len(venny1), \"cells; \", genes[1], \": expressed in \", len(venny2), \"cells. \")\n                \n\n        if len(genes) == 3:\n            venny3 = meds.loc[genes[2]]\n            venny3 = venny3[venny3 >= 0.5].index.to_list()\n            if (len(venny1) >= 1) & (len(venny2) >= 1) & (len(venny3) >= 1):\n                name3 = str(genes[2]) + \" only\"\n                namebc = str(genes[1]) + \" and \" + str(genes[2])\n                nameac = str(genes[0]) + \" and \" + str(genes[2])\n                nameabc = str(genes[0]) + \" and \" + str(genes[1])+ \" and \" + str(genes[2])         \n                v3only = list(set(venny1) - set(venny2) - set(venny3))\n                overlap_bc = [value for value in venny2 if value in venny3]\n                overlap_ac = [value for value in venny1 if value in venny3]\n                overlap_abc = [value for value in venny3 if value in overlap_ab]\n\n                la,lb,lc, ld, le, lf, lg  = len(v1only),len(v2only),len(v3only),len(overlap_ab),len(overlap_bc),len(overlap_ac),len(overlap_abc)\n                # now find the max\n                max_len = max(la,lb,lc,ld, le, lf, lg)\n                if not max_len == la:\n                    v1only.extend(['']*(max_len-la))\n                if not max_len == lb:\n                    v2only.extend(['']*(max_len-lb))\n                if not max_len == lc:\n                    v3only.extend(['']*(max_len-lc))\n                if not max_len == ld:\n                    overlap_ab.extend(['']*(max_len-ld))\n                if not max_len == le:\n                    overlap_bc.extend(['']*(max_len-le))\n                if not max_len == lf:\n                    overlap_ac.extend(['']*(max_len-lf))\n                if not max_len == lg:\n                    overlap_abc.extend(['']*(max_len-lg))\n\n                venn_df = pd.DataFrame({name1:v1only,name2:v2only,name3:v3only,nameab:overlap_ab,namebc:overlap_bc,\n                                                nameac:overlap_ac, nameabc: overlap_abc}) \n                        \n                labels = [\"Cells containing \" + genes[0],\"Cells containing \" \n                                + genes[1], \"Cells containing \" +genes[2] ]\n                v = venn3([set(venny1), set(venny2), set(venny3)], \n                                set_labels = (genes[0], genes[1], genes[2]))\n                        \n            else:\n                print(\"No Overlap: At least one gene is not expressed over a value of 0.5 in any cell types:\", \n                      genes[0], \": expressed in \", len(venny1), \"cells; \", genes[1], \": expressed in \", len(venny2), \"cells; \",\n                     genes[2], \": expressed in \", len(venny3), \"cells.\")\n        \n        plt.show()","execution_count":null,"outputs":[]},{"metadata":{"hide_input":false,"trusted":true},"cell_type":"code","source":"### cell #1\n\nbox_layout = widgets.Layout(\n        border='solid 1px black',\n        margin='0px 10px 10px 0px',\n        padding='5px 5px 5px 5px',\n        height='auto', width='auto')\n\ngrid = GridspecLayout(1, 3)\ngrid.layout = box_layout\n# fill it in with widgets\ngrid[0,0] = specieschoices\ngrid[0,1] = spacerhuman\ngrid[0, 2:] = graphs2\ngrid","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"name":"python3","display_name":"Python 3","language":"python"},"language_info":{"name":"python","version":"3.7.6","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat":4,"nbformat_minor":4}