{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Stress Microarray Grapher:\n",
    "    This program allows you to filter the three microarrays comparing PACAP Knockout, Pac1 Knockout, and Control mice following restraint stress. First, set your threshold for p-value and fold-change. Pressing the \"Set Threholds\" button will make the gene selection pane appear. The genes available in the drop-down are only those that reached your previously established significance level in at least one comparison. Choose each gene in the dropdown and click \"Enter Gene\". WHen you have selected your genes, click \"Produce Output.\" \n",
    "    \n",
    "    \n",
    "    NOTE- Reset is not working very well now- if you want to "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "hideCode": false,
    "hidePrompt": true
   },
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "import pandas as pd\n",
    "pd.set_option('mode.chained_assignment', None)\n",
    "import ipywidgets as widgets\n",
    "import io\n",
    "import plotly.express as px\n",
    "from IPython.display import display\n",
    "from ipywidgets import interact_manual, interactive,GridspecLayout, AppLayout, Button, Layout, Widget"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "hideCode": false,
    "hidePrompt": true
   },
   "outputs": [],
   "source": [
    "##DATA FOR STRESS COMPARISONS\n",
    "#retreive file with ANOVA data\n",
    "result = pd.read_csv(r\"C:\\Users\\bakalarda\\Desktop\\Bioinformatics Rapgef2\\EidenLab Microarray\\for stress v nostress\\Data\\Processed Expression Matrixes\\statistical_results.csv\")\n",
    "#load the expression data and combine/ fix older ones\n",
    "a=pd.read_csv(r\"C:\\Users\\bakalarda\\Desktop\\Bioinformatics Rapgef2\\EidenLab Microarray\\for stress v nostress\\Data\\Processed Expression Matrixes\\expressions_2019.04 3hr restraint hypothalamus part1.csv\")\n",
    "b=pd.read_csv(r\"C:\\Users\\bakalarda\\Desktop\\Bioinformatics Rapgef2\\EidenLab Microarray\\for stress v nostress\\Data\\Processed Expression Matrixes\\expressions_2019.04 3hr restraint hypothalamus part2.csv\")\n",
    "c=pd.read_csv(r\"C:\\Users\\bakalarda\\Desktop\\Bioinformatics Rapgef2\\EidenLab Microarray\\for stress v nostress\\Data\\Processed Expression Matrixes\\expressions_2019.04 3hr restraint adrenal part1.csv\")\n",
    "d=pd.read_csv(r\"C:\\Users\\bakalarda\\Desktop\\Bioinformatics Rapgef2\\EidenLab Microarray\\for stress v nostress\\Data\\Processed Expression Matrixes\\expressions_2019.04 3hr restraint adrenal part2.csv\")\n",
    "e=pd.read_csv(r\"C:\\Users\\bakalarda\\Desktop\\Bioinformatics Rapgef2\\EidenLab Microarray\\for stress v nostress\\Data\\Processed Expression Matrixes\\expressions_2019.03 1hr restraint hypothalamus part1.csv\")\n",
    "f=pd.read_csv(r\"C:\\Users\\bakalarda\\Desktop\\Bioinformatics Rapgef2\\EidenLab Microarray\\for stress v nostress\\Data\\Processed Expression Matrixes\\expressions_2019.03 1hr restraint hypothalamus part2.csv\")\n",
    "frames =[a,b,c,d,e,f]\n",
    "expressions = pd.concat(frames, sort=False).reset_index()\n",
    "expressions.drop(expressions.columns[0], axis=1, inplace=True)\n",
    "expressions.drop(expressions.columns[0], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "hideCode": false,
    "hidePrompt": true
   },
   "outputs": [],
   "source": [
    "##define GUI \n",
    "     \n",
    "def choose_gene(change):\n",
    "    if dropdown.value not in genes2:\n",
    "        if dropdown.value != '':\n",
    "            genes2.append(dropdown.value)\n",
    "            with outgenes:\n",
    "                display(dropdown.value)\n",
    "                \n",
    "graph_out1 = widgets.Output()  \n",
    "graph_out2 = widgets.Output()\n",
    "app_out = widgets.Output()\n",
    "\n",
    "\n",
    "def on_value_change(change):\n",
    "    #just be here\n",
    "    print('')\n",
    "\n",
    "    \n",
    "            \n",
    "def on_reset_button_clicked(reset_button):\n",
    "    app_out.clear_output()\n",
    "    graph_out1.clear_output() \n",
    "    graph_out2.clear_output()\n",
    "    outerrors.clear_output()\n",
    "    outgenes.clear_output()\n",
    "    outgraph.clear_output()\n",
    "\n",
    "    \n",
    "def on_go_button_clicked(go_button):\n",
    "    mouse_means2, stats, sig_genes = get_vals(go_button)\n",
    "#Mark errors\n",
    "    if len(genes2) < 1:\n",
    "        with outerrors:\n",
    "            print(\"Choose at least one Gene\")\n",
    "    else:\n",
    "\n",
    "       ###Graph expression   \n",
    "        import plotly.graph_objects as go\n",
    "        with graph_out1:\n",
    "\n",
    "            fig = px.bar(mouse_means2, x=\"combo\", y=\"average\",\n",
    "                        facet_row = \"Gene Symbol\",\n",
    "                        color = \"combo\",\n",
    "                        facet_col = \"Experiment\",\n",
    "                        hover_data = [\"Gene Symbol\",\"average\",\"Tissue\"],\n",
    "                        color_discrete_sequence= px.colors.sequential.Plasma,\n",
    "                        error_y=\"error\",\n",
    "\n",
    "                  ).for_each_trace(lambda t: t.update(name=t.name.split(\"=\")[1]))\n",
    "\n",
    "            for a in fig.layout.annotations:\n",
    "                        a.text = a.text.split(\"=\")[1]\n",
    "                        fig.update_yaxes(title_text='')\n",
    "                        fig.update_xaxes(title_text='')\n",
    "                        fig.update_layout(\n",
    "                            title=\"Expression levels of selected genes in groups where threshold qualifications were met\",\n",
    "                            #xaxis_title=\"Group\",\n",
    "                            yaxis_title=\"Fold-change\",\n",
    "                        xaxis={'categoryorder':'total descending'})\n",
    "\n",
    "            fig.show()\n",
    "\n",
    "            ################################################################\n",
    "        with graph_out2:\n",
    "            fig = px.bar(stats, x=\"Comparison\", y=\"Fold-change\",\n",
    "                       # width=800, height=500,\n",
    "                        facet_col = \"Gene Symbol\",\n",
    "                        color_discrete_sequence= px.colors.sequential.Plasma,\n",
    "                        color = \"p-value\", range_color=[0,p_threshold.value], #Change so user input tells it what p and fc to use\n",
    "                        hover_name = \"Experiment\",\n",
    "                        hover_data = [\"Tissue\", \"Fold-change\", \"p-value\"]\n",
    "                    )\n",
    "                        #.for_each_trace(lambda t: t.update(name=t.name.split(\"=\")[1]))\n",
    "            fig.for_each_annotation(lambda a: a.update(text=a.text.split(\"=\")[-1]))\n",
    "            fig.update_layout(title=\"Fold-changes and p-values\",\n",
    "                             )\n",
    "\n",
    "            fig.show()\n",
    "\n",
    "def on_thresh_button_clicked(th_go_button):\n",
    "    \n",
    "    #first clear any existing stuff out\n",
    "    \n",
    "    graph_out1.clear_output() \n",
    "    graph_out2.clear_output()\n",
    "    outerrors.clear_output()\n",
    "    outgenes.clear_output()\n",
    "    outgraph.clear_output()\n",
    "    genes2=[]\n",
    "    results_stringent = result[result[\"p-value\"] <= p_threshold.value]#Change so user input tells it what p and fc to use\n",
    "    results_stringent2=results_stringent[(results_stringent['Fold-change']>= fc_threshold.value) | (results_stringent['Fold-change']<= -fc_threshold.value)]\n",
    "    sig_genes0 = results_stringent2[\"Gene Symbol\"].tolist()\n",
    "    sig_genes = list(set(sig_genes0))\n",
    "    mouse_means1 = expressions[expressions[\"Gene Symbol\"].isin(genes2)]\n",
    "    mouse_means1['combo']=mouse_means1['Genotype'] + \" \" + mouse_means1['Stress Condition']\n",
    "    mouse_means2 = (mouse_means1.groupby(['Experiment','Gene Symbol','Tissue','combo'])['Expression']\n",
    "         .agg([('average','mean'),('error','sem')])\n",
    "         .reset_index())\n",
    "    stats = results_stringent2[results_stringent2[\"Gene Symbol\"].isin(genes2)]\n",
    "\n",
    "    #create next part of app\n",
    "    dropdown.options = sig_genes\n",
    "    with app_out:\n",
    "        display(app)\n",
    "    return(mouse_means2, stats)\n",
    "\n",
    "\n",
    "def get_vals(go_button):\n",
    "    results_stringent = result[result[\"p-value\"] <= p_threshold.value]#Change so user input tells it what p and fc to use\n",
    "    results_stringent2=results_stringent[(results_stringent['Fold-change']>= fc_threshold.value) | (results_stringent['Fold-change']<= fc_threshold.value)]\n",
    "    sig_genes = results_stringent2[\"Gene Symbol\"].tolist()\n",
    "    mouse_means1 = expressions[expressions[\"Gene Symbol\"].isin(genes2)]\n",
    "    mouse_means1['combo']=mouse_means1['Genotype'] + \" \" + mouse_means1['Stress Condition']\n",
    "    mouse_means2 = (mouse_means1.groupby(['Experiment','Gene Symbol','Tissue','combo'])['Expression']\n",
    "         .agg([('average','mean'),('error','sem')])\n",
    "         .reset_index())\n",
    "    stats = results_stringent2[results_stringent2[\"Gene Symbol\"].isin(genes2)]\n",
    "\n",
    "    return(mouse_means2, stats, sig_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "hideCode": false,
    "hidePrompt": true,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "68dcbb967b98444da23500e48b2290d8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(Label(value='First, set p and fold-change'), BoundedFloatText(value=0.05, description='p-value:â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "########################################################################################################\n",
    "#App Prep\n",
    "style = {'description_width': 'initial'}\n",
    "genes2 = []       \n",
    "spacer = widgets.Label('')\n",
    "genetitle = widgets.Label('Select Genes')\n",
    "\n",
    "dropdown =  widgets.Combobox(\n",
    "    options= [\"sig_genes\"],\n",
    "    description='Gene',\n",
    "    placeholder='Type here',\n",
    "    layout={'width': '80%'})\n",
    "\n",
    "   \n",
    "outgenes = widgets.Output()   \n",
    "outgraph = widgets.Output(layout={'border': '1px solid black'})  \n",
    "\n",
    "        \n",
    "gene_button = widgets.Button(description = 'Enter Gene', \n",
    "                button_style = '', \n",
    "                display='flex',\n",
    "                flex_flow='column',\n",
    "                align_items='stretch')\n",
    "gene_button.on_click(choose_gene)\n",
    "\n",
    "go_button = widgets.Button(description = 'Produce Output', \n",
    "                           button_style = '', \n",
    "                            display='flex',\n",
    "                            flex_flow='column',\n",
    "                            align_items='stretch')\n",
    "go_button.on_click(on_go_button_clicked)\n",
    "\n",
    "reset_button = widgets.Button(description = 'Reset App', \n",
    "                           button_style = '', \n",
    "                            display='flex',\n",
    "                            flex_flow='column',\n",
    "                            align_items='stretch')\n",
    "reset_button.on_click(on_reset_button_clicked)\n",
    "\n",
    "\n",
    "gene_button.observe(dropdown.value)\n",
    "go_button.observe(dropdown.value)\n",
    "\n",
    "p_threshold = widgets.BoundedFloatText(\n",
    "    value=0.05,\n",
    "    description='p-value:',\n",
    "    disabled=False,\n",
    "    width = '80%',\n",
    "    min=0,\n",
    "    max=0.1,\n",
    "    step=0.01,\n",
    "    style=style)\n",
    "p_threshold.observe(on_value_change, names='value')\n",
    "\n",
    "fc_threshold = widgets.BoundedFloatText(\n",
    "    value=2,\n",
    "    description='Fold-change (+/-):',\n",
    "    disabled=False,\n",
    "    min=1.5,\n",
    "    max=20,\n",
    "    step=0.5,\n",
    "    width = '80%',\n",
    "    style=style)\n",
    "fc_threshold.observe(on_value_change, names='value')\n",
    "\n",
    "\n",
    "graph_out1 = widgets.Output()  \n",
    "graph_out2 = widgets.Output() \n",
    "\n",
    "##########################################################################\n",
    "##Deploy the app\n",
    "\n",
    "threshold_title = widgets.Label('First, set p and fold-change')\n",
    "th_go_button = widgets.Button(description = 'Set Thresholds', \n",
    "                           button_style = '', \n",
    "                            display='flex',\n",
    "                            flex_flow='column',\n",
    "                            align_items='stretch',\n",
    "                                   value = \"no\")\n",
    "th_go_button.on_click(on_thresh_button_clicked)\n",
    "genes_out_title = widgets.Label('Genes Chosen:')\n",
    "\n",
    "\n",
    "genesbox = widgets.VBox([genetitle,dropdown])\n",
    "genes_out = widgets.VBox([spacer,genes_out_title, outgenes])\n",
    "buttons = widgets.HBox([gene_button, go_button, reset_button])\n",
    "#cellsbox = widgets.VBox([celllabel, Cells_drop])\n",
    "genechoices = widgets.HBox([genesbox, genes_out])\n",
    "\n",
    "outerrors = widgets.Output() \n",
    "\n",
    "graphit = widgets.VBox([graph_out1, graph_out2])\n",
    "top_app = widgets.VBox([genechoices, outerrors, buttons]) ##Put seletcors for p and fc values here\n",
    "app = widgets.VBox([top_app, graphit])\n",
    "thresholds = widgets.VBox([threshold_title,p_threshold, fc_threshold, th_go_button,app_out])\n",
    "\n",
    "thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Hide code",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
