{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "hideCode": true,
    "hideOutput": true,
    "hidePrompt": true,
    "hide_input": true
   },
   "source": [
    "# Press the double arrow above and enjoy! This program allows you to examine gene expression in all cell clusters identified in Loo, L., Simon, J. M., Xing, L., McCoy, E. S., Niehaus, J. K., Guo, J., ... & Zylka, M. J. (2019). Single-cell transcriptomic analysis of mouse neocortical development. Nature communications, 10(1), 1-11.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hideCode": true,
    "hideOutput": true,
    "hidePrompt": true
   },
   "source": [
    "# Cells are grouped into Cell Types (Inhibitory, Cortical Layers, Non-neuronal, Proliferative, Midbrain). They are further divided into Clusters (identified by Loo et al) and age of sample. Each 'block' on each bar is an individual cluster. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "hideCode": true,
    "hideOutput": true,
    "hidePrompt": true,
    "hide_input": false
   },
   "outputs": [],
   "source": [
    "import matplotlib \n",
    "import ipywidgets as widgets\n",
    "import pandas as pd\n",
    "import io\n",
    "import seaborn as sns\n",
    "from IPython.display import display\n",
    "from ipywidgets import interact_manual, interactive,GridspecLayout, AppLayout, Button, Layout, Widget\n",
    "import plotly.express as px\n",
    "\n",
    "data = pd.read_csv(\"LooSimon_MouseCortexDropSeq_E14P0_avgExpByCellType.csv\")\n",
    "# Get column names \n",
    "listOfColumnNames = list(data.columns.values[1:])\n",
    "idx = list(data['GeneID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "hideCode": true,
    "hideOutput": true,
    "hidePrompt": true
   },
   "outputs": [],
   "source": [
    "cells =[]\n",
    "genes = []\n",
    "\n",
    "def on_go_button_clicked(go_button):\n",
    "\n",
    "#Mark errors\n",
    "    if len(genes) < 1:\n",
    "        with outerrors:\n",
    "            print(\"Choose at least one Gene\")\n",
    "    if len(list(Cells_drop.value)) < 1:\n",
    "        with outerrors:\n",
    "            print(\"Choose at least one Cell Cluster\")\n",
    "            \n",
    "## restructure data frame to include selected types only \n",
    "    if len(genes) >= 1:\n",
    "         if len(list(Cells_drop.value)) >= 1:\n",
    "            data_gene_cell = data[list(Cells_drop.value)]\n",
    "            data_gene_cell.index= data['GeneID']\n",
    "            data_gene_cell = data_gene_cell.loc[genes]\n",
    "            data_gene_cell = data_gene_cell.reset_index()\n",
    "\n",
    "        ## reshape for plotting\n",
    "            data_gene_cell2 = pd.melt(data_gene_cell, id_vars =['GeneID'],\n",
    "                              var_name ='Cluster', value_name ='Expression')\n",
    "\n",
    "        ## plot\n",
    "            ages = []\n",
    "            cell_type = []\n",
    "            supergroup = []\n",
    "            import re\n",
    "\n",
    "            for i in data_gene_cell2['Cluster']:    \n",
    "                if '-P]' in i:\n",
    "                    ages.append('p0')\n",
    "                if '-E]' in i:\n",
    "                    ages.append('e14.5')\n",
    "                if 'Layer' in i:\n",
    "                    supergroup.append(\"Cortical Layers\")\n",
    "                if any(x in i for x in ['Endothelial','Choroid','Astrocyte','Oligo','Ganglionic','Micro']): \n",
    "                    supergroup.append(\"Non-Neuronal\")\n",
    "                if any(x in i for x in ['SVZ','RG']): \n",
    "                    supergroup.append(\"Proliferative\")\n",
    "                if \"Int\" in i:\n",
    "                    supergroup.append(\"Inhibitory\")\n",
    "                if any(x in i for x in ['Thalamic','Striatal']): \n",
    "                    supergroup.append(\"Midbrain\")\n",
    "                j = re.sub(\"[\\[].*?[\\]]\", \"\", i)\n",
    "                cell_type.append(j)\n",
    "\n",
    "            data_gene_cell2['Cell_Type'] = cell_type\n",
    "            data_gene_cell2['Age'] = ages\n",
    "            data_gene_cell2['Category'] = supergroup\n",
    "\n",
    "\n",
    "            import plotly.graph_objects as go\n",
    "            fig = px.bar(data_gene_cell2, x=\"Category\", y=\"Expression\", color = \"Age\", \n",
    "                    facet_col = \"GeneID\",\n",
    "                    hover_name = \"Cell_Type\",\n",
    "                    hover_data = [\"Age\", \"Cluster\"]\n",
    "                    )\n",
    "            fig.update_layout(barmode='group')\n",
    "            with graph_out1:\n",
    "                    fig.show()\n",
    "\n",
    "            ## print table\n",
    "            fig2 = go.Figure(data=[go.Table(\n",
    "            header=dict(values=list(data_gene_cell2.columns),\n",
    "                    fill_color='paleturquoise',\n",
    "                    align='left'),\n",
    "            cells=dict(values=[data_gene_cell2.GeneID, data_gene_cell2.Cluster, data_gene_cell2.Expression,\n",
    "                              data_gene_cell2.Cell_Type, data_gene_cell2.Age, data_gene_cell2.Category],\n",
    "                    fill_color='lavender',\n",
    "                    align='left'))\n",
    "                ])\n",
    "            with graph_out2:\n",
    "                    fig2.show()\n",
    "\n",
    "\n",
    "def choose_gene(change):\n",
    "    if dropdown.value not in genes:\n",
    "        if dropdown.value != '':\n",
    "            genes.append(dropdown.value)\n",
    "            with outgenes:\n",
    "                display(dropdown.value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "hideCode": true,
    "hideOutput": true,
    "hidePrompt": true,
    "hide_input": false
   },
   "outputs": [],
   "source": [
    "##define GUI  \n",
    "\n",
    "spacer = widgets.Label('')\n",
    "genetitle = widgets.Label('Select Genes')\n",
    "\n",
    "dropdown =  widgets.Combobox(\n",
    "    options= idx,\n",
    "    description='Gene',\n",
    "    placeholder='Type here',\n",
    "    layout={'width': '80%'})\n",
    "\n",
    "   \n",
    "outgenes = widgets.Output()   \n",
    "outgraph = widgets.Output(layout={'border': '1px solid black'})  \n",
    "\n",
    "        \n",
    "gene_button = widgets.Button(description = 'Enter Gene', \n",
    "                button_style = '', \n",
    "                display='flex',\n",
    "                flex_flow='column',\n",
    "                align_items='stretch')\n",
    "\n",
    "gene_button.on_click(choose_gene)\n",
    "\n",
    "celllabel = widgets.Label('Cell Clusters:')\n",
    "Cells_drop =widgets.SelectMultiple(\n",
    "    options=listOfColumnNames,\n",
    "    rows=8,\n",
    "    layout={'width': 'max-content'}) \n",
    "\n",
    "\n",
    "go_button = widgets.Button(description = 'Produce Output', \n",
    "                           button_style = '', \n",
    "                            display='flex',\n",
    "                            flex_flow='column',\n",
    "                            align_items='stretch')\n",
    "go_button.on_click(on_go_button_clicked)\n",
    "\n",
    "gene_button.observe(dropdown.value)\n",
    "go_button.observe(dropdown.value)\n",
    "go_button.observe(Cells_drop.value)\n",
    "\n",
    "\n",
    "graph_out1 = widgets.Output()  \n",
    "graph_out2 = widgets.Output()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hideCode": true,
    "hidePrompt": true,
    "hide_input": false
   },
   "outputs": [],
   "source": [
    "genesbox = widgets.VBox([genetitle,dropdown, gene_button])\n",
    "cellsbox = widgets.VBox([celllabel, Cells_drop])\n",
    "genechoices = widgets.VBox([genesbox, outgenes,go_button])\n",
    "outerrors = widgets.Output() \n",
    "\n",
    "graphit = widgets.VBox([graph_out1, graph_out2])\n",
    "top_app = widgets.HBox([cellsbox,genechoices, outerrors])\n",
    "app = widgets.VBox([top_app, graphit])\n",
    "\n",
    "\n",
    "app"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
